Implement append and split_off for VecMap (RFC 509)

msiemens opened the pull request:
Implements `append()` and `split_off()` for `VecMap`. It's worth noting that `append()` will overwrite existing keys (the RFC doesn't specify how `append()` should handle duplicate keys).

cc #19986

r? @Gankro

Gankro commented on src/libcollections/vec_map.rs:
````diff
@@ -318,6 +318,79 @@ impl<V> VecMap<V> {
+    /// assert_eq!(b[4], "d");
+    /// ```
+    pub fn split_off(&mut self, at: usize) -> Self {
+        assert!(at <= self.len(), "`at` out of bounds");
````
For Sorted Maps, `at` shouldn't have any assert, I reckon. It's basically a predecessor query.

Gankro replied:
That is, the backing storage is an implementation detail for Maps, unlike sequences.

Gankro commented on src/libcollections/vec_map.rs:
```diff
@@ -318,6 +318,79 @@ impl<V> VecMap<V> {
+        other.v.extend((0..at).map(|_| None));
+
+        // Move elements beginning with `at` from `self` into `other`
+        other.v.extend(self.v.split_off(at).into_iter());
```
This unnecessarily allocates a vector, only to discard it.

Gankro replied:
I would maybe do something like `self.v[at...].iter_mut().map(|opt| opt.take())`, although you need to check that `at` isn't OOB.

Gankro replied:
Hmm no this will do more work than necessary too.

Gankro commented on src/libcollections/vec_map.rs:
```diff
@@ -1140,6 +1213,52 @@ mod test_map {
+        let mut b = VecMap::new();
+        b.insert(3, "d");  // Overwrite element from a
+        b.insert(4, "e");
+        b.insert(5, "f");
```
For reference, you can collect an iterator of tuples. e.g. [(1, "a"), (2, "b"), (3, "c")].

Gankro commented on src/libcollections/vec_map.rs:
```diff
@@ -318,6 +318,79 @@ impl<V> VecMap<V> {
+        let mut other = VecMap::new();
+
+        // First fill the new VecMap with `None`s until `at`
+        other.v.extend((0..at).map(|_| None));
```
If I have the map `[(1, 1), (2, 2), (3, 3)]`, and you `split_off(10000000)` this will be pretty catastrophic.

Gankro replied:
Basically I think the "right" approach is to defer doing anything until you find an element in `self` that actually will get placed in `other`. Only once you have done this should you fill up the Vec with None's from the last insertion point to the current point (Vec::extend is your friend here, I imagine).

Otherwise you could just ignore the backing storage and use VecMap::insert directly. Not sure how the perf compares.

msiemens replied:
So it would first get the index of the first non-`None` item, fill up `other` with `None` to this index and then proceed extending with the rest of `self.v`?

msiemens replied:
Also note that the latest commit replaced the `assert!` with a check whether `a > self.len()` in order to return an empty VecMap in that case. Thus, `split_off(100000)` would be short-circuited

msiemens replied:
I've now updated the PR with the approach you described

Gankro commented:
I'm not sure exactly how much VecMap should try to help out with foot-guns. In particular should it assume that inputs are reasonable and that VecMaps are small and dense? Or should it make a best effort not to do something catastrophic when the VecMap has a lot of empty space (particularly at the tail).

CC @huonw

Gankro commented on src/libcollections/vec_map.rs:
```diff
@@ -318,6 +318,92 @@ impl<V> VecMap<V> {
+
+        if at == 0 {
+            // Move all elements to other
+            other.append(self);
```
This seems crazy inefficient. Why not just mem::swap(self, other)?

Gankro commented on src/libcollections/vec_map.rs:
```diff
@@ -318,6 +318,92 @@ impl<V> VecMap<V> {
+        other.v.extend((0..start_index).map(|_| None));
+
+        // Move elements beginning with `start_index` from `self` into `other`
+        other.v.extend(self.v.iter_mut().skip(start_index).map(|el| el.take()));
```
Rather than skip, consider `self.v[start_index..].iter_mut().map()`. 

One-time cost vs constant branching.

Gankro commented:
r=me with the last two changes and squashed commits. :smile:

msiemens commented:
I've addressed both comments and squashed the commits. I think, we're ready to go :tada: .
Thanks for reviewing, @Gankro, I'm learning a lot from your feedback!

msiemens commented:
By the way, should't I also include a `#[unstable(...)]` tag for both methods?

Gankro commented:
Oh, whoops, yes! I am still getting used to the new stability stuff.

msiemens commented:
I've added them now. BTW: Is there a convention when to use `#[inline]`?

Gankro commented:
Ugh #[inline] is basically cargo culting in the standard libraries IMHO. It does two things: It hints to LLVM that inlining is more desriable than usual, and it makes the function a candidate for cross-crate-inling (by including the source in the compiled crate).

The first one is *very* dubious, because LLVM's defaults *should* be smarter than us in this regard. The second one is completely inapplicable for almost all collections code because *generic* code is *already* available for cross-crate inlining (since you need the source for monomorphization).

It's a battle I simply don't have the energy to fight. Everyone does it. It'd be great if you excluded them IMO, but I can't fight the world on this.

msiemens commented:
> It'd be great if you excluded them IMO, but I can't fight the world on this.

Alright, I've removed them. I always thought that the compiler should know better what to inline. Didn't know about the cross-crate-inlining, though.

Gankro commented:
@bors r+ ec7145e

Great work!

bors commented:
:hourglass: Testing commit ec7145e with merge 81cec05...

bors commented:
:broken_heart: Test failed - [auto-mac-64-opt](http://buildbot.rust-lang.org/builders/auto-mac-64-opt/builds/3753)

Manishearth commented:
```
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/libcollections/vec_map.rs:364:5: 366:6 error: method is never used: `append`, #[deny(dead_code)] on by default
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/libcollections/vec_map.rs:364     fn append(&mut self, other: &mut Self) {
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/libcollections/vec_map.rs:365         self.extend(other.drain());
/Users/rustbuild/src/rust-buildbot/slave/auto-mac-64-opt/build/src/libcollections/vec_map.rs:366     }
```

Gankro commented:
Whoops, needs a `pub`.

msiemens commented:
Oops! Fixed

Manishearth commented:
@bors: r=Gankro 25d3e01
